### PR TITLE
Add zstd support for llvm 16

### PR DESCRIPTION
**Pull request by SaltyKitkat**
###### Description of changes

LLVM add zstd support since 16

And maybe we should support that.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
Can you detail more what is meant by zstd support and what does it bring concretely?

**Comment by rrbutani** [comment]
> Can you detail more what is meant by zstd support and what does it bring concretely?

@RaitoBezarius This enables features like [`clang -gz=zstd`](https://github.com/llvm/llvm-project/commit/a0365abad8113d7e60294c5e581c24ec06ba037e) (compressing debug sections with zstd) and corresponding [support](https://reviews.llvm.org/D133548) in [`lld`](https://github.com/llvm/llvm-project/commit/449f2ca146dde0ef14c9837f71d4a0be92ca4843) and in the LLVM binutils.

**Review by rrbutani — CHANGES_REQUESTED** [review]
Thanks for opening this PR! `zstd` support in LLVM is definitely something we overlooked.

---

A few comments:
  - `zstd` support in `clang` and `lld` did land in LLVM 16 but `zstd` was added to the LLVM support library in LLVM 15: https://github.com/llvm/llvm-project/commit/e939bf67e34037970192fa90ad22ab7628f5de33; can you amend this PR to backport this change to `llvmPackages_15` as well?
    + edit: will do in a follow up PR
  - [ ] it'd be nice to have some tests for this (i.e. a passthru test on `clang` that attempts to pass `-gz=zstd` if `enableZstd` is `true` on the underlying `llvm`) but I don't think it's a blocker for this PR
  - [ ] LLVM also has a [`LLVM_USE_STATIC_ZSTD` option](https://github.com/llvm/llvm-project/commit/fc1da043f4f9198303abd6f643cf23439115ce73) (added LLVM 16) that does not appear to be autodetected; we may want to pass this explicitly as part of the cmake flags when `stdenv.hostPlatform.isStatic` is `true` (though I am not sure if this is actually required...)

File: pkgs/development/compilers/llvm/16/lldb/default.nix [hunk 1]
```diff
@@ -76,6 +77,7 @@ stdenv.mkDerivation (rec {
   buildInputs = [
     ncurses
     zlib
+    zstd
```
[the remark this answers is not in the document]
Comment (SaltyKitkat) [reply pkgs/development/compilers/llvm/16/lldb/default.nix:80]: I remove both zlib and zstd in this file
and it seems that lldb can still be compiled with zlib and zstd support?
cmake found zlib and zstd anyway.

File: pkgs/development/compilers/llvm/16/lldb/default.nix [hunk 2]
```diff
@@ -4,6 +4,7 @@
 , cmake
 , ninja
 , zlib
+, zstd
```
[the remark this answers is not in the document]
Comment (SaltyKitkat) [reply pkgs/development/compilers/llvm/16/lldb/default.nix:7]: > Is there some functionality in `lldb` that requires `zstd` as a dep here?

Actually I don't know. I just search where `zlib` appears, and add `zstd`. LOL

**Comment by SaltyKitkat** [comment]
@rrbutani oops… backporting this to llvm15 seems cause lots of rebuild

should we open a separate pr for that?

**Comment by rrbutani** [comment]
> backporting this to llvm15 seems cause lots of rebuild
> 
> should we open a separate pr for that?

Sure! If/when you open that PR, be sure to target `staging`.

File: pkgs/development/compilers/llvm/16/llvm/default.nix [hunk 3]
```diff
@@ -96,7 +97,7 @@ in
   buildInputs = [ libxml2 libffi ]
     ++ optional enablePFM libpfm; # exegesis
 
-  propagatedBuildInputs = [ ncurses zlib ];
+  propagatedBuildInputs = [ ncurses zlib zstd ];
```
[the remark this answers is not in the document]
Comment (SaltyKitkat) [reply pkgs/development/compilers/llvm/16/llvm/default.nix:100]: > Can we gate this on an `enableZstd` input to the package (defaulted to `true)?

I'm a little confused here.

- Why we need a feature gate here? It seems that no zstd support brings no benefits from my perspective?

- I know nothing about tests in nixpkgs. Maybe I need some help about the tests.

[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply pkgs/development/compilers/llvm/16/llvm/default.nix:100]: Feature gate is useful to control closure sizes and sometimes when something make a special variant incompatible, e.g. static builds, musl builds, cross compilation, etc.

[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply pkgs/development/compilers/llvm/16/llvm/default.nix:100]: If we passthru it, we can recognize whether any LLVM, Clang, LLD at hand is able to do zstd or not and we can decide whether our test harness will attempt to decompress stuff using zstd. Implementing the passthru is the only thing needed here at the moment AFAIK.

**Comment by rrbutani** [comment]
Now that #236420 has reached `master` I think we'll need to retarget this PR to `staging`; there are too many LLVM 16 rdeps (via Rust 1.70).

@SaltyKitkat would you mind adding back in the LLVM 15 changes and retargeting this PR? (instructions for retargeting to `staging` are [here](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging))

**Comment by SaltyKitkat** [comment]
Oops…

**Pull request closed without merging**